### ISSUE TITLE
Web Server HTTP Header Memory Exhaustion DoS in HTTPServer

### ISSUE THREAD

**Issue by Drsuperman** [prometheus/client_java#718]
It was possible to kill the web server by sending an invalid 'infinite' HTTP request that never ends, like:
GET / HTTP/1.0 Referer: XXXXXXXXXXXXXXXXXXXXXXXX ...

An attacker may exploit this vulnerability to make your web server crash continually (if the attack saturates virtual memory on the target) or even execute arbitrary code on your system (in case of buffer / heap overflow).

**Comment by fstab** [comment]
I failed to reproduce this with OpenJDK 11 and the example application in `./integration_tests/example_application/`.

Here's what I did:

* Start a Docker container with the example application and a memory limit of 500m
  ```
  ./mvnw package
  cd ./integration_tests/example_application/
  docker run -m 500m --rm -p 9000:9000 -v $(pwd)/target:/target -ti openjdk:11  java -jar /target/example_application.jar
  ```
* Send a request like you posted above to port 9000
  ```
  ( echo -n 'GET / HTTP/1.0 Referer: ' ; while true ; do echo -n X ; done ) | nc localhost 9000
  ```
* Run a few regular requests to check if the example application still works
  ```
  curl localhost:9000/metrics
  ```
Result: After a while the large request fails with `Ncat: Broken pipe.` while the example application is still doing fine and responding to regular requests. Maybe the error happens only with a specific Java version?
